presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/CrossJoinsElimination.java

```diff
@@ -0,0 +1,194 @@
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Set;
+
+import static com.facebook.presto.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class CrossJoinsElimination
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            Map<Symbol, Type> types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator)
+    {
+        if (!SystemSessionProperties.isJoinReorderingEnabled(session)) {
+            return plan;
+        }
+
+        List<JoinGraph> joinGraphs = JoinGraph.buildFrom(plan);
+
+        for (int i = 0; i < joinGraphs.size(); i++) {
+            JoinGraph graph = joinGraphs.get(i);
+            if (!graph.isOriginalPlanWithCrossJoin()) {
+                // If there was no cross join in original plan do not reorder joins
+                continue;
+            }
+
+            Optional<List<Integer>> joinOrder = getJoinOrder(graph);
+
+            if (joinOrder.isPresent()) {
+                plan = rewriteWith(new Rewriter(idAllocator, graph, joinOrder.get()), plan);
+                joinGraphs = JoinGraph.buildFrom(plan);
+            }
+        }
+
+        return plan;
+    }
+
+    /**
+     * Given JoinGraph determine the order of joins between graph nodes
+     * by traversing JoinGraph. Any graph traversal algorithm could be used
+     * here (like BFS or DFS), but we use PriorityQueue to preserve
+     * original JoinOrder as mush as it is possible. PriorityQueue returns
+     * next nodes to join in order of their occurrence in original Plan.
+     */
+    public static Optional<List<Integer>> getJoinOrder(JoinGraph graph)
+    {
+        ImmutableList.Builder<PlanNode> joinOrder = ImmutableList.builder();
+
+        Map<PlanNode, Integer> priorities = new HashMap<>();
+        for (int i = 0; i < graph.size(); i++) {
+            priorities.put(graph.getNode(i), i);
+        }
+
+        PriorityQueue<PlanNode> nodesToVisit = new PriorityQueue<>(
+                graph.size(),
+                (Comparator<PlanNode>) (node1, node2) -> priorities.get(node1).compareTo(priorities.get(node2)));
+        Set<PlanNode> visited = new HashSet<>();
+
+        nodesToVisit.add(graph.getNode(0));
+
+        while (!nodesToVisit.isEmpty()) {
+            PlanNode node = nodesToVisit.poll();
+            if (visited.contains(node)) {
+                continue;
+            }
+            visited.add(node);
+            joinOrder.add(node);
+            for (JoinGraph.Edge edge : graph.getEdges(node)) {
+                PlanNode neighbour = edge.getTargetNode();
+                nodesToVisit.add(neighbour);
+            }
+        }
+
+        if (visited.size() != graph.size()) {
```

Comment:
why do we abort reordering in this case.  If there is a query that contains some actual cross join we still want to eliminate unnecessary cross joins from the subgraphs